api/octopusApi.go

```diff
@@ -150,42 +151,163 @@ func (a *Api) SetHandlers(prefix string, rtr *mux.Router) {
-	if groupId, err := a.authorizeAndGetGroupId(res, req, vars); err != nil {
+
+	start := time.Now()
+
+	if a.authorized(req) {
+		userId := vars["userID"]
+		groupId, err := a.getUserPairId(userId, a.getToken(req))
+		if err != nil {
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("TimeLastEntryUserAndDevice: failed after [%.5f] secs with error[%s]", time.Now().Sub(start).Seconds(), err.Error()))
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+		if a.userCanViewData(userId, groupId) {
+			timeLastEntry, err := a.Store.GetTimeLastEntryUserAndDevice(groupId, vars["deviceID"])
+			if err != nil {
+				log.Println(QUERY_API_PREFIX, fmt.Sprintf("TimeLastEntryUserAndDevice: failed after [%.5f] secs", time.Now().Sub(start).Seconds()))
+				log.Println(QUERY_API_PREFIX, "TimeLastEntryUserAndDevice:", error_running_query, err.Error())
+				http.Error(res, error_running_query, http.StatusInternalServerError)
+				return
+			}
+
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("TimeLastEntryUserAndDevice: completed in [%.5f] secs", time.Now().Sub(start).Seconds()))
+			res.Header().Set("content-type", "application/json")
+			res.WriteHeader(http.StatusOK)
+			res.Write(timeLastEntry)
+			return
+		}
+		log.Println(QUERY_API_PREFIX, fmt.Sprintf("TimeLastEntryUserAndDevice: failed after [%.5f] secs with error[%s]", time.Now().Sub(start).Seconds(), error_no_permissons))
+		http.Error(res, error_no_permissons, http.StatusForbidden)
 		return
-	} else {
+	}
+	log.Println(QUERY_API_PREFIX, fmt.Sprintf("TimeLastEntryUserAndDevice: failed authorization after [%.5f] secs", time.Now().Sub(start).Seconds()))
+	http.Error(res, "failed authorization", http.StatusUnauthorized)
+	return
+
+}
+
+// http.StatusOK - the requested data
+// http.StatusBadRequest - something was wrong with the request data
+// http.StatusUnauthorized - you don't have a valid token
+func (a *Api) Query(res http.ResponseWriter, req *http.Request) {
+
+	start := time.Now()
+
+	if a.authorized(req) {
+
+		log.Println(QUERY_API_PREFIX, "Query: starting ... ")
 
-		deviceId := vars["deviceID"]
+		defer req.Body.Close()
+		rawQuery, err := ioutil.ReadAll(req.Body)
 
-		timeLastEntry := a.Store.GetTimeLastEntryUserAndDevice(groupId, deviceId)
+		if err != nil || string(rawQuery) == "" {
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("Query: err decoding nonempty response body: [%v]\n [%v]\n", err, req.Body))
+			http.Error(res, error_building_query, http.StatusBadRequest)
+			return
+		}
+		query := string(rawQuery)
+
+		log.Println(QUERY_API_PREFIX, "Query: raw ", query)
+
+		errs, qd := model.BuildQuery(query)
+
+		if len(errs) != 0 {
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("Query: errors [%v] found parsing raw query [%s]", errs, query))
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("Query: failed after [%.5f] secs", time.Now().Sub(start).Seconds()))
+			http.Error(res, fmt.Sprintf("Errors building query: [%v]", errs), http.StatusBadRequest)
```

Comment:
Might want to put the errs array out as JSON rather than Go format for parsability?